File: src/main/java/com/mcommings/campaigner/modules/calendar/entities/Event.java
package com.mcommings.campaigner.modules.calendar.entities;

import com.mcommings.campaigner.modules.common.entities.Campaign;
import com.mcommings.campaigner.modules.locations.entities.City;
import com.mcommings.campaigner.modules.locations.entities.Continent;
import com.mcommings.campaigner.modules.locations.entities.Country;
import jakarta.persistence.*;
import lombok.*;

@Getter
@Setter
@Builder
@Entity
@Table(name = "events",
        uniqueConstraints = {
                @UniqueConstraint(
                        columnNames = {"name", "fk_campaign_uuid"}
                )
        })
@NoArgsConstructor
@AllArgsConstructor
public class Event {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(nullable = false, unique = true)
    private String name;

    @Column(nullable = false)
    private String description;

    @Column(name = "event_year")
    private int year;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_month", nullable = false)
    private Month month;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_week", nullable = false)
    private Week week;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_day", nullable = false)
    private Day day;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_continent", nullable = false)
    private Continent continent;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_country", nullable = false)
    private Country country;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_city", nullable = false)
    private City city;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_campaign_uuid", nullable = false)
    private Campaign campaign;
}

File: src/main/frontend/lib/models/calendar/event.dart
class Event {
  final int id;
  final String name;
  final String description;
  final int eventYear;
  final int? fkMonth;
  final int? fkWeek;
  final int? fkDay;
  final int? fkContinent;
  final int? fkCountry;
  final int? fkCity;
  final String fkCampaignUuid;

  const Event({
    required this.id,
    required this.name,
    required this.description,
    required this.eventYear,
    this.fkMonth,
    this.fkWeek,
    this.fkDay,
    this.fkContinent,
    this.fkCountry,
    this.fkCity,
    required this.fkCampaignUuid,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    try {
      return Event(
        id: json['id'] as int,
        name: json['name'] as String,
        description: json['description'] as String,
        // backend uses snake_case for event year in some places; accept either
        eventYear: (json['event_year'] ?? json['eventYear']) as int,
        fkMonth: json['fk_month'] as int?,
        fkWeek: json['fk_week'] as int?,
        fkDay: json['fk_day'] as int?,
        fkContinent: json['fk_continent'] as int?,
        fkCountry: json['fk_country'] as int?,
        fkCity: json['fk_city'] as int?,
        fkCampaignUuid: json['fk_campaign_uuid'] as String,
      );
    } catch (e) {
      throw FormatException('Event parse error: $e');
    }
  }
}
